### ISSUE TITLE
Datatable with lazyDataModel lost selections on pagination

### ISSUE THREAD

**Issue by Nikoas83** [primefaces/primefaces#1833]
## 1) Environment
- PrimeFaces version: 5.3
- Does it work on the newest PrimeFaces version? I can't test with 6.0 (and earlier 5.X) but it seems this is not in bugFixes of release notes
- Application server + version: Widlfly 9.0.2
- Affected browsers: All

All are explained on this topic : [http://forum.primefaces.org/viewtopic.php?f=3&t=47224&p=147483#p147483](url)
## 2) Expected behavior

In datatable with lazy datamodel, multiselections with checkboxes and pagination, when selects rows in first page, and after, click on other page then return in first page, i want to see selected items checked (and highlighted).
## 3) Actual behavior

When selects rows in first page, after click on other page then go back to first page, selected items are not checked (and not highlighted), but selected list stay well up to date in backing bean.
## 4) Steps to reproduce

Just create lazy datatable with multiselection, implements well lazyDataModel, and enable pagination.
## 5) Sample XHTML

See [http://forum.primefaces.org/viewtopic.php?f=3&t=47224&p=147483#p147483](url)

`            <h:form id="usersManagerForm" >

```
            <p:dataTable id="usersTable" var="user" value="#{usersTestManagerBean.lazyDataModel}" lazy="true" widgetVar="users_table"
                  selection="#{usersTestManagerBean.selectedComptesList}"
                  paginator="true" rows="5" paginatorAlwaysVisible="true" paginatorPosition="bottom"
                  rowsPerPageTemplate="5,20,50,100,500" > 

                <p:ajax event="rowSelect" update=":usersManagerForm:usersTable:usersSelection" />
                <p:ajax event="rowSelectCheckbox" global="false" immediate="true" update=":usersManagerForm:usersTable:usersSelection" />
                <p:ajax event="rowUnselectCheckbox" global="false" immediate="true"  update=":usersManagerForm:usersTable:usersSelection" />
                <p:ajax event="toggleSelect" global="false" immediate="true"  update=":usersManagerForm:usersTable:usersSelection" /> 

                <f:facet name="header">
                    <h:panelGroup id="usersSelection" >
                        <h:outputText value="#{not empty usersTestManagerBean.selectedComptesList ? ''.concat(usersTestManagerBean.selectedComptesList.size()).concat(' items selected') : '0 item selected'}" />
                    </h:panelGroup>
                </f:facet>

               <!-- Checkbox selection -->
               <p:column selectionMode="multiple" styleClass="colonne-checkbox" />

               <!-- Group -->
               <p:column  > 
                       <f:facet name="header">
                           <h:outputText value="#{msg.Nom_famille}" />
                       </f:facet>
                   <h:outputText value="#{user.nom}" />
               </p:column>

               <!-- Name -->
               <p:column  > 
                   <f:facet name="header">
                           <h:outputText value="#{msg.Prenom}" />
                       </f:facet>
                   <h:outputText value="#{user.prenom}" /> 
               </p:column>

               <!-- Login -->
               <p:column > 
                   <f:facet name="header">
                           <h:outputText value="#{msg.Login}" />
                       </f:facet>              
                   <h:outputText value="#{user.login}" />   
               </p:column>     

            </p:dataTable>

         </h:form>`
```
## 6) Sample bean

See [http://forum.primefaces.org/viewtopic.php?f=3&t=47224&p=147483#p147483](url)

ManagedBean 
`@ViewScoped
public class UsersTestManagerBean implements Serializable {

```
private static final long serialVersionUID = 1L;
private List<Compte> selectedComptesList;
private LazyDataModel<Compte> compteLazyDataModel;


@PostConstruct
public void init() {
    selectedComptesList = null;
    compteLazyDataModel = new UsersTestLazyDataModel();
}

public LazyDataModel<Compte> getLazyDataModel() {
    return this.compteLazyDataModel;
}

public List<Compte> getSelectedComptesList() {
    return selectedComptesList;
}

public void setSelectedComptesList(List<Compte> selectedComptesList) {
    this.selectedComptesList = selectedComptesList;
}

public void unselectComptesList() {
    selectedComptesList = null;
}

public int getSizeOfSelectedComptesList() {
    return selectedComptesList.size();
}
```

}`

LazyDataModel
`public class UsersTestLazyDataModel extends LazyDataModel<Compte> {

```
private static final long serialVersionUID = 1L;
List<Compte> compteList;
List<Compte> compteListFull;

public UsersTestLazyDataModel() {
    compteListFull = new ArrayList<>();
    for(int i=0; i < 20; i++) {
        compteListFull.add(buildNewRdmCompte(i));
    }
}

@Override
public List<Compte> load(int first, int pageSize, String sortField, SortOrder sortOrder, Map<String, Object> filters) {

    // Do not handle filters and sorting, not displayed, it is not necessary for test
    compteList = compteListFull.subList(first, first + pageSize);
    this.setRowCount(compteListFull.size());

    return compteList;
}

// Build new user, Compte is simple object
private Compte buildNewRdmCompte(int i) {
    Compte c = new Compte();
    c.setId(i);
    c.setPrenom("Name" + i);
    c.setNom("Surname" + i);
    c.setLogin("Login" + i);
    return c;
}


@Override
public Compte getRowData(String infoId) {
    Integer id = Integer.valueOf(infoId);

    for (Compte account : compteList) {
        if (id.equals(account.getId())) {
            return account;
        }
    }
    return null;
}


@Override
public Object getRowKey(Compte compte) {
    return compte.getId();
}
```

}`

**Comment by kukel** [comment]
Not a bug as can be seen in http://forum.primefaces.org/viewtopic.php?f=3&t=47224

**Comment by Mitsch79** [comment]
Really? I'd definately consider this being a bug and it's still present in primefaces 6.1. Given the fact, that eager (non-lazy) tables behave exactly the way OP expects it's not conclusive that a lazy table suddenly should not. Plus, the lazy table highlights (checks) the previously selected rows as soon as any row on the corresponding page is selected. So, the view still knows about the formerly selected items but just doesn't check them upon navigating (paginating) back to the page the selection was made on.

The argument "it's not common to have selections across the pages" is not valid in my opinion since the functionality is there but it's not working correctly. And if i'd want to have no multi page selection i'd simply implement a page listener that clear's the selection (at my will ;). The hint to use the ajax listeners is a workaround for this bug, but in architectural and design mannors... well it's a workaround ;) As well as the approach of implementing an empty setSelectedItems method.

**Comment by Rapster** [comment]
I'm facing this problem in 6.2, but seems to be fixed in 6.3 @Nikoas83 @Mitsch79 can you have a try?

**Comment by Betlista** [comment]
When no ajax events are used it works fine in 6.2 (I shared my code in forum - https://forum.primefaces.org/viewtopic.php?f=3&t=57687), consequence is, that you do not have list of selected items available (at least I' not aware of), but on submit it is available.

**Comment by Rapster** [comment]
@Betlista not related to this problem, wait for an answer on the forum (or use stackoverflow)

**Comment by Betlista** [comment]
@Rapster How is that not related, when it seems to me working in 6.2? You said it is not working in 6.2 and in forum it is marked as resolved for 5.3 too...

I specifically mean from expected behaviour part:
> When selects rows in first page, after click on other page then go back to first page, selected items are not checked (and not highlighted), but selected list stay well up to date in backing bean.

This is not the case (based on my testing in 6.2), when ajax events are not used, which is not mentioned as needed, from my point of view it is simply wrong implemenation... New bug should be opened for ajax events support of multiselect datatable (or it might help you solving it).
